doc/specs/#888 - Com Api.md

```diff
@@ -0,0 +1,889 @@
+    [contract(Microsoft.Management.Deployment.WindowsPackageManagerContract, 1)]
+    runtimeclass CatalogPackage
+    {
+        /// IMPLEMENTATION NOTE: This would be new work as Windows Package Manager does not currently create a 
+        /// CatalogPackage object from a manifest.
+        /// DESIGN NOTE: This is proposed as the preferred alternative to having InstallOptions match the winget
+        /// command line which allows you to pass in a manifestPath directly into install. Allowing creation of 
+        /// a CatalogPackage object from a manifest allows the install api to consistently use package objects 
+        /// rather than having a "you can either set the manifest path or use a package but not both" problem.
+        static CatalogPackage TryCreateFromManifest(String manifestPath);
+
+        /// IMPLEMENTATION NOTE: PackageProperty fields from AppInstallerRepositorySearch.h
+        /// Gets a property of this package.
+        String Id { get; };
+        String Name { get; };
+
+        /// Gets the installed package information if the package is installed.
+        PackageVersionInfo InstalledVersion{ get; };
+
+        /// Gets all available versions of this package.
+        /// The versions will be returned in sorted, descending order.
+        ///  Ex. { 4, 3, 2, 1 }
+        Windows.Foundation.Collections.IVectorView<PackageVersionId> AvailableVersions { get; };
+
+        /// Gets a specific version of this package.
+        PackageVersionInfo LatestAvailableVersion { get; };
+
+        /// Gets a specific version of this package.
+        PackageVersionInfo GetAvailableVersion(PackageVersionId versionKey);
+
+        /// Gets a value indicating whether an available version is newer than the installed version.
+        Boolean IsUpdateAvailable { get; };
+
+        /// Gets whether the package is installing
+        Boolean IsInstalling { get; };
+
+        /// DESIGN NOTE:
+        /// IsSame from IPackage in AppInstallerRepositorySearch is not implemented in V1.
+        /// Determines if the given IPackage refers to the same package as this one.
+        /// virtual bool IsSame(const IPackage*) const = 0;
+    }
+
+    /// IMPLEMENTATION NOTE: CompositeSearchBehavior from AppInstallerRepositorySource.h
+    /// Search behavior for composite sources.
+    /// Only relevant for composite sources with a local source, not for aggregates of multiple available sources.
+    /// Installed and available packages in the result are always correlated when possible.
+    [contract(Microsoft.Management.Deployment.WindowsPackageManagerContract, 1)]
+    enum CompositeSearchBehavior
+    {
+        /// Search only installed packages
+        InstalledPackages,
+        /// Search only installing packages
+        InstallingPackages,
+        /// Search only local packages (installed and installing)
+        AllLocalPackages,

```

Comment:
What about packages that are uninstalling? Or queued but not yet downloaded? Or downloaded and not staged? Or staged but not yet begun installing? Do they also count as "All"?

Reply:
The plan for V1 is No, Yes, Yes, Yes. I could definitely imagine adding an UninstallingPackages local catalog though, so yes, perhaps this should just be a flags enum rather than a defined All.

Reply:
DISCUSSION: This enables asking about which packages are installed.  Composite catalogs have many child catalogs. The caller might want to specify which kind of catalog to look at.  The question is to return a package that's available but not installed yet.

RECOMMEND: Narrow to three options - installed, available-local, available-remote.

CONSIDER: Don't merge local and remote catalogs for search; have developers search individual catalogs.